code/controllers/subsystem/tts.dm

```diff
@@ -0,0 +1,226 @@
+		// If it took too long to process, don't bother playing it
+		var/timeout_time = current_message[START_TIME_INDEX] + message_timeout
+		if(response.errored || timeout_time < world.time)
+			cached_voices -= current_message[IDENTIFIER_INDEX]
+			continue
+
+		var/identifier = current_message[IDENTIFIER_INDEX]
+		var/sound/new_sound = new("tmp/tts/[identifier].ogg")
+		play_tts(current_message[TARGET_INDEX], new_sound, current_message[LANGUAGE_INDEX], current_message[LOCAL_INDEX], current_message[MESSAGE_RANGE_INDEX])
+		for(var/extra_target in current_message[EXTRA_TARGETS_INDEX])
+			play_tts(extra_target["target"], new_sound, current_message[LANGUAGE_INDEX], extra_target["local"], extra_target["range"])
+		cached_voices -= identifier
+		if(MC_TICK_CHECK)
+			return
+
+/datum/controller/subsystem/tts/proc/queue_tts_message(target, message, datum/language/language, speaker, filter, local = FALSE, message_range = 7)
+	if(!tts_enabled)
+		return
+
+	var/static/regex/contains_alphanumeric = regex("\[a-zA-Z0-9]", "g")
+	// If there is no alphanumeric char, the output will usually be static, so
+	// don't bother sending
+	if(contains_alphanumeric.Find(message) == 0)
+		return
+
+	var/shell_scrubbed_input = tts_alphanumeric_filter(message)
+	shell_scrubbed_input = copytext(shell_scrubbed_input, 1, 300)
+	var/identifier = sha1(speaker + shell_scrubbed_input + filter)
+	var/cached_voice = cached_voices[identifier]
+	if(islist(cached_voice))
+		cached_voice[EXTRA_TARGETS_INDEX] += list(list("target" = target, "local" = local, "range" = message_range))
+		return
+	else if(fexists("tmp/tts/[identifier].ogg"))
+		var/sound/new_sound = new("tmp/tts/[identifier].ogg")
+		play_tts(target, new_sound, language, local, message_range)
+		return
+	if(!(speaker in available_speakers))
+		return
+	speaker = tts_alphanumeric_filter(speaker)
+
+	var/list/headers = list()
+	headers["Content-Type"] = "application/json"
+	headers["Authorization"] = CONFIG_GET(string/tts_http_token)
+	var/datum/http_request/request = new()
+	var/file_name = "tmp/tts/[identifier].ogg"
+	request.prepare(RUSTG_HTTP_METHOD_GET, "[CONFIG_GET(string/tts_http_url)]/tts?voice=[speaker]&identifier=[identifier]&filter=[url_encode(filter)]", json_encode(list("text" = shell_scrubbed_input)), headers, file_name)
+	var/list/waiting_list = queued_tts_messages
+	if(length(in_process_tts_messages) < max_concurrent_requests)
+		request.begin_async()
+		waiting_list = in_process_tts_messages
+
+	var/list/data = list(
+		// TARGET_INDEX = 1
+		target,
+		// IDENTIFIER_INDEX = 2
+		identifier,
+		// START_TIME_INDEX = 3
+		world.time,
+		// REQUEST_INDEX = 4
+		request,
+		// MESSAGE_INDEX = 5
+		shell_scrubbed_input,
+		// EXTRA_TARGETS_INDEX = 6
+		list(),
+		// LANGUAGE_INDEX = 7
+		language,
+		// LOCAL_INDEX = 8
+		local,
+		// MESSAGE_RANGE_INDEX = 9
+		message_range
+	)
+	cached_voices[identifier] = data
+	waiting_list += list(data)
+	sortTim(waiting_list, GLOBAL_PROC_REF(cmp_word_length_asc))
```

Comment:
i feel like we used to have a priority queue datum

Reply:
If there is one, I can't find it

Reply:
we have /datum/heap